pkg/controller/servingcert/controller/secret_creating_controller.go

```diff
@@ -0,0 +1,325 @@
+		serviceCopy.Annotations[api.ServingCertErrorNumAnnotation] = strconv.Itoa(getNumFailures(serviceCopy) + 1)
+		_, updateErr := sc.serviceClient.Services(serviceCopy.Namespace).Update(serviceCopy)
+
+		// if we're past the max retries and we successfully updated, then the sync loop successfully handled this service and we want to forget it
+		if updateErr == nil && getNumFailures(serviceCopy) >= sc.maxRetries {
+			return nil
+		}
+		return err
+	}
+	if kapierrors.IsAlreadyExists(err) {
+		actualSecret, err := sc.secretClient.Secrets(serviceCopy.Namespace).Get(secret.Name, metav1.GetOptions{})
+		if err != nil {
+			// if we have an error creating the secret, then try to update the service with that information.  If it fails,
+			// then we'll just try again later on  re-list or because the service had already been updated and we'll get triggered again.
+			serviceCopy.Annotations[api.ServingCertErrorAnnotation] = err.Error()
+			serviceCopy.Annotations[api.ServingCertErrorNumAnnotation] = strconv.Itoa(getNumFailures(serviceCopy) + 1)
+			_, updateErr := sc.serviceClient.Services(serviceCopy.Namespace).Update(serviceCopy)
+
+			// if we're past the max retries and we successfully updated, then the sync loop successfully handled this service and we want to forget it
+			if updateErr == nil && getNumFailures(serviceCopy) >= sc.maxRetries {
+				return nil
+			}
+			return err
+		}
+
+		if actualSecret.Annotations[api.ServiceUIDAnnotation] != string(serviceCopy.UID) {
+			serviceCopy.Annotations[api.ServingCertErrorAnnotation] = fmt.Sprintf("secret/%v references serviceUID %v, which does not match %v", actualSecret.Name, actualSecret.Annotations[api.ServiceUIDAnnotation], serviceCopy.UID)
+			serviceCopy.Annotations[api.ServingCertErrorNumAnnotation] = strconv.Itoa(getNumFailures(serviceCopy) + 1)
+			_, updateErr := sc.serviceClient.Services(serviceCopy.Namespace).Update(serviceCopy)
+
+			// if we're past the max retries and we successfully updated, then the sync loop successfully handled this service and we want to forget it
+			if updateErr == nil && getNumFailures(serviceCopy) >= sc.maxRetries {
+				return nil
+			}
+			return errors.New(serviceCopy.Annotations[api.ServingCertErrorAnnotation])
+		}
+	}
+
+	serviceCopy.Annotations[api.ServingCertCreatedByAnnotation] = sc.commonName()
+	delete(serviceCopy.Annotations, api.ServingCertErrorAnnotation)
+	delete(serviceCopy.Annotations, api.ServingCertErrorNumAnnotation)
+	_, err = sc.serviceClient.Services(serviceCopy.Namespace).Update(serviceCopy)
+
+	return err
+}
+
+func getNumFailures(service *v1.Service) int {
+	numFailuresString := service.Annotations[api.ServingCertErrorNumAnnotation]
+	if len(numFailuresString) == 0 {
+		return 0
+	}
+
+	numFailures, err := strconv.Atoi(numFailuresString)
+	if err != nil {
+		return 0
+	}
+
+	return numFailures
+}
+
+func (sc *ServiceServingCertController) requiresCertGeneration(service *v1.Service) bool {
```

Comment:
this function should be kept to its original logic

Reply:
Fixed